lib/ui/viewport_canvas.dart

```diff
@@ -0,0 +1,511 @@
+      ..width = '0'
+      ..height = '0'
+      ..overflow = 'hidden';
+    _element.add(_content);
+
+    final spacers = <CoreElement>[
+      _spaceTop,
+      _spaceBottom,
+      _spaceLeft,
+      _spaceRight
+    ];
+    for (var spacer in spacers) {
+      spacer.element.style
+        ..position = 'absolute'
+        ..overflow = 'hidden';
+    }
+    _spaceLeft.element.style
+      ..top = '0'
+      ..bottom = '0'
+      ..left = '0'
+      ..width = '0';
+    _spaceRight.element.style
+      ..top = '0'
+      ..bottom = '0'
+      ..right = '0'
+      ..width = '0';
+    _spaceTop.element.style
+      ..top = '0'
+      ..left = '0'
+      ..right = '0'
+      ..height = '0';
+    _spaceBottom.element.style
+      ..bottom = '0'
+      ..left = '0'
+      ..right = '0'
+      ..height = '0';
+
+    _content.add(spacers);
+
+    // TODO(jacobr): remove call to allowInterop once
+    // https://github.com/dart-lang/sdk/issues/35484 is fixed.
+    _visibilityObserver = IntersectionObserver(
+        allowInterop(_onVisibilityChange), {'root': _element.element});
+    for (var spacer in spacers) {
+      _visibilityObserver.observe(spacer.element);
+    }
+
+    element.onScroll.listen((_) {
+      if (_currentMouseHover != null) {
+        _dispatchMouseMoveEvent();
+      }
+      // Add a buffer when mouse wheel scrolling as that event is unfortunately
+      // async so we risk flickering UI if we don't render with a buffer.
+      rebuild(force: false);
+    });
+    if (_onTap != null) {
+      _content.onClick.listen((e) {
+        _onTap(_clientToGlobal(e.client));
+      });
+    }
+    _content.element.onMouseLeave.listen((_) {
+      _currentMouseHover = null;
+      if (_onMouseLeave != null) {
+        _onMouseLeave();
+      }
+    });
+    _content.element.onMouseMove.listen((e) {
+      _currentMouseHover = e.client;
+      _dispatchMouseMoveEvent();
+    });
+  }
+
+  void _dispatchMouseMoveEvent() {
+    if (_onMouseMove != null) {
+      _onMouseMove(_clientToGlobal(_currentMouseHover));
+    }
+  }
+
+  Offset _clientToGlobal(Point client) {
+    final elementRect = _content.element.getBoundingClientRect();
+    return Offset(client.x - elementRect.left, client.y - elementRect.top);
+  }
+
+  Point _currentMouseHover;
+
+  /// Id used to help debug what was rendered as part of the current frame.
+  int _frameId = 0;
+
+  final CanvasPaintCallback _paintCallback;
+  final MouseCallback _onTap;
+  final MouseCallback _onMouseMove;
+  final VoidCallback _onMouseLeave;
+  final Map<_ChunkPosition, _CanvasChunk> _chunks = {};
+
+  static const int maxChunks = 50;
+
+  /// Intersection observer used to detect when
+  IntersectionObserver _visibilityObserver;
+
+  CoreElement get element => _element;
+  final CoreElement _element;
+
+  /// Element containing all content that scrolls within the viewport.
+  final CoreElement _content;
+
+  // Fixed sized divs placed around the the content that is currently rendered
```

Comment:
nit: I would add a period after rendered to break this up a bit and make it more readable.

Reply:
done